torch/testing/_internal/common_device_type.py

```diff
@@ -618,25 +564,91 @@ class OpDTypes(Enum):
 #     operator's gradient formula supports
 #   OpDTypes.unsupported_backward - the test is instantiated for all dtypes the
 #     operator's gradient formula doesn't support
-#   OpDTypes.none - the test is instantied without any dtype. The dtype
-#     arg will be set to _NO_DTYPES.
+#   OpDTypes.none - the test is instantied without any dtype. The test signature
+#     should not include a dtype kwarg in this case.
 #
 # These options allow tests to have considerable control over the dtypes
 #   they're instantiated for. Finally, the @dtypes decorator composes with the
 #   @ops decorator, and works the same as the "dtypes" argument to @ops.
 
-class ops(object):
+class ops(_TestParameterizer):
     def __init__(self, op_list, *, dtypes: OpDTypes = OpDTypes.basic,
                  allowed_dtypes: Optional[Sequence[torch.dtype]] = None):
         self.op_list = op_list
         self.opinfo_dtypes = dtypes
         self.allowed_dtypes = set(allowed_dtypes) if allowed_dtypes is not None else None
 
-    def __call__(self, fn):
-        fn.op_list = self.op_list
-        fn.allowed_dtypes = self.allowed_dtypes
-        fn.opinfo_dtypes = self.opinfo_dtypes
-        return fn
+    def _parameterize_test(self, test, generic_cls, device_cls):
+        """ Parameterizes the given test function across each op and its associated dtypes. """
+        for op in self.op_list:
+            # Acquires dtypes, using the op data if unspecified
+            dtypes = device_cls._get_dtypes(test)
+            if dtypes is None:
+                if self.opinfo_dtypes == OpDTypes.unsupported_backward:
+                    dtypes = set(get_all_dtypes()).difference(op.supported_backward_dtypes(device_cls.device_type))
+                elif self.opinfo_dtypes == OpDTypes.supported_backward:
+                    dtypes = op.supported_backward_dtypes(device_cls.device_type)
+                elif self.opinfo_dtypes == OpDTypes.unsupported:
+                    dtypes = set(get_all_dtypes()).difference(op.supported_dtypes(device_cls.device_type))
+                elif self.opinfo_dtypes == OpDTypes.supported:
+                    dtypes = op.supported_dtypes(device_cls.device_type)
+                elif self.opinfo_dtypes == OpDTypes.basic:
+                    dtypes = op.default_test_dtypes(device_cls.device_type)
+                elif self.opinfo_dtypes == OpDTypes.none:
+                    dtypes = [None]
+                else:
+                    raise RuntimeError(f"Unknown OpDType: {self.opinfo_dtypes}")
+
+                if self.allowed_dtypes is not None:
+                    dtypes = dtypes.intersection(self.allowed_dtypes)
+            else:
+                assert self.allowed_dtypes is None, "ops(allowed_dtypes=[...]) and the dtypes decorator are incompatible"
```

Comment:
While we're here I think this assert is redundant with the assert that follows it?

Reply:
Hm, I think you can hit the second assert with:
```
@ops(op_db, allowed_dtypes=None, dtypes=OpDTypes.supported)
@dtypes(torch.float32)
def test_something(self, device, dtype, op):
    ...
```

Reply:
Good point